Pipeline: Simplify wrapper

ansonwong9695 opened the pull request:
- add ```Simplify``` binary to ```third_party```
- add its wrapper in ```wrappers```
- add ```SIMPLIFICATION_RATIO``` in env as default value
- add simplify calls to all pipelines

UdomkarnBoonyaprasert committed:
PIPELINE: add simplify wrapper
add wrapper and its binary executable(not sure where i should put it)

UdomkarnBoonyaprasert committed:
PIPELINE: add chmod in simplify wrapper

UdomkarnBoonyaprasert committed:
PIPELINE: fix bugs in simplify wrapper
- remove flags from subprocess run
- change data type in argument

UdomkarnBoonyaprasert committed:
PIPELINE: add env variable to wrapper
add wrapper calls to bone and lung pipeline in POST_PROCESSING(left abdominal pipeline for now)

UdomkarnBoonyaprasert committed:
Merge branch 'PIPELINE/major-refactoring' of https://github.com/nbckr/HoloRepository-Core into PIPELINE/simplify_wrapper

UdomkarnBoonyaprasert committed:
PIPELINE: add simplification to abdomen pipeline

UdomkarnBoonyaprasert committed:
PIPELINE: update .env

ansonwong9695 committed:
PIPELINE: remove chmod from wrapper

ansonwong9695 committed:
PIPELINE: moved 'simplify' binary to third_party

nielsboecker commented on HoloPipelines/core/wrappers/simplify.py:
```diff
@@ -0,0 +1,29 @@
+
+from config import SIMPLIFICATION_RATIO
+
+current_script_directory = (
```
replace with `simplify_binary_path = "./core/third_party/fast_quadric_mesh_simplification/simplify"`

nielsboecker commented on HoloPipelines/core/wrappers/simplify.py:
```diff
@@ -0,0 +1,29 @@
+
+from config import SIMPLIFICATION_RATIO
+
+current_script_directory = (
```
Can you move the binary to that subdirectory and create a small README in the directory too please, stating where it is taken from. Like [>this<](https://github.com/nbckr/HoloRepository-Core/blob/PIPELINE/major-refactoring/HoloPipelines/core/third_party/ct_lung_segmentation/README.md).

UdomkarnBoonyaprasert replied:
moved and ```README.md``` added

nielsboecker commented on HoloPipelines/core/wrappers/simplify.py:
```diff
@@ -0,0 +1,29 @@
+    obj_output_path: str,
+    simplification_ratio: float = SIMPLIFICATION_RATIO,
+):
+    # NOTE: link to simplify repository https://github.com/sp4cerat/Fast-Quadric-Mesh-Simplification
```
Remove from here

nielsboecker commented on HoloPipelines/core/wrappers/simplify.py:
```diff
@@ -0,0 +1,29 @@
+):
+    # NOTE: link to simplify repository https://github.com/sp4cerat/Fast-Quadric-Mesh-Simplification
+    simplify_command = [
+        "./simplify",
```
Replace with `simplify_binary_path`

nielsboecker commented on HoloPipelines/core/wrappers/simplify.py:
```diff
@@ -0,0 +1,29 @@
+        str(simplification_ratio),
+    ]
+
+    completed_process = subprocess.run(simplify_command, cwd=current_script_directory)
```
Remove the `cwd`

nielsboecker reviewed (commented):
Code looks pretty good! Please look at my comments.

Also can you please add a simple test? It's sufficient to just call the method and verify that a file exists in the output path afterwards. You may want to find a very small obj to include in the `utils` directory as input for that purpose.

nbckr commented:
Btw @ansonwong9695, @UdomkarnBoonyaprasert I merged this locally with the tests branch and the 3 pipeline tests fail. I think it's because the path to the `temp.obj` file is broken due to the way you use `cwd`. So I suspect that once you follow my PR suggestions, it will also fix that bug.

nielsboecker reviewed (changes requested):
Paths are broken (again)

nbckr commented:
**TODO**: Once my comments are sorted out, change PR target to `dev`

UdomkarnBoonyaprasert committed:
PIPELINE: address comments from PR
- update path to simplify
- moved Simplify binary
- added README for Simplify
- added testing

UdomkarnBoonyaprasert committed:
PIPELINE: removed comment

UdomkarnBoonyaprasert commented:
I added a sample obj file (```tests/utils/teddy.obj```) for testing. It's around 90 KB, would this be ok?

nbckr commented:
Yes that sounds like a good idea. But currently, the test fails, maybe due to absolute vs relative path to test data.

UdomkarnBoonyaprasert committed:
PIPELINE: simplify wrapper test fix
I'm not certain if this will fix it, as I can't use this binary on my OS. New fix will be up soon if test fails.

nielsboecker committed:
PIPELINE: Refactor test_simplify_wrapper

nielsboecker committed:
Merge branch 'dev' into PIPELINE/simplify_wrapper

nielsboecker reviewed (approved):
Thanks @UdomkarnBoonyaprasert I have made some refactorings and I am happy with this PR now. @boonwj Going to leave it up to you to do the final merge, as you are currently deploying the system.

I have verified that the bone + lung pipeline are still functional locally. Haven't tried dockerised versions nor the abdominal pipeline, but it should be fine, as it is just the additional step with the simplify wrapper.

nielsboecker committed:
PIPELINE: Update import in test_nifti_file

boonwj commented:
The different workflows do appear to work in docker. There's just a new configuration field (SIMPLIFICATION_RATIO) that I need to take care of when I deploy it to kubes.

Also, there does not seem to be any documentation on the individual configuration fields. Pipeline's readme and the configuration section only mentioned that .env is the configuration file. 

If there's time, it will be nice to document the individual fields.
